---
layout: page
title: About
permalink: /about/
---

MBLâ€™s Workshop on Molecular Evolution is the most prestigious workshop serving the field of evolutionary studies. Founded in 1988, it is the longest-running workshop of its kind, and it has earned worldwide recognition for its rich and intensive learning experience. Students work closely with internationally-recognized scientists, receiving 
* high-level instruction in the principles of molecular evolution and evolutionary genomics
* advanced training in statistical methods best suited to modern datasets, and 
* hands-on experience with the latest software tools (often from the authors of the programs they are using).

The material is delivered via lectures, discussions, and bioinformatic exercises motivated by contemporary topics in molecular evolution. A hallmark of this workshop is the direct interaction between students and field-leading scientists. The workshop serves graduate students, postdocs, and established faculty from around the world seeking to apply the principles of molecular evolution to questions of anthropology, conservation genetics, development, behavior, physiology, and ecology. The workshop also welcomes participants from federal agencies and science journalists. A priority of this workshop is to foster an environment where students can learn from each other as well from the course faculty.

Content has been carefully selected to provide participants with the background and practical skills required by modern molecular datasets. The schedule addresses the following subject areas, with each subject having one or more exercises focused on practical data analysis and interpretation skills.

* An evolutionary perspective on molecular data: protein sequence versus protein structure; homology, orthology and parology; multiple sequence alignment; information resources
* Foundations of phylogenetic analysis: theoretical, mathematical, and statistical principles; sampling properties of sequence data; Maximum likelihood theory and practice; Bayesian analysis; hypothesis testing
* Species-level phylogenomics: species trees from gene trees; species delimitation; multi locus and SNP data; empirical examples
* Deep phylogenomics: deep evolutionary relationships; lateral gene transfer; modeling approaches; topology testing; sequencing strategies
* Foundations of population genetic analysis: neutral theory; coalescence theory; maximum likelihood and Bayesian estimation of population genetic parameters; empirical examples
* Population genomics: phylogeography; molecular ecology; next-generation population genetics; signatures of natural selection; natural populations of non-model organisms
* Comparative genomics: genome content; genome structure; gene and genome evolutionary dynamics; prediction of gene function
* Molecular evolution integrated at organism and higher levels: population biology and ecology; natural selection; systematics and conservation
* Molecular evolution integrated at lower levels: biochemistry; cell biology; physiology; natural selection; relationship of genotype to phenotype

As the course progresses, participants learn how to use the following software to address questions concerning the origins, maintenance, and function of molecular variation: PAUP*, IQTree, SVD Quartets, ASTRAL, SNaQ, RevBayes, MIGRATE, PAML, and other tools for data preparation and interpreting results. Students will work with software on their own laptops and learn how to run analyses on a high performance computer cluster. Training in programming with specific examples in phylogenetics is provided too.

