---
layout: page
title: Joe Bielawski
permalink: /faculty-bielawski/
---
{% include figure.html description="Joe Bielawski" url="/faculty/bielawski/img/JPB.jpg" css="image-right" height="250px" %}
## Institutional home pages

* [Research group page](http://www.bielawski.info)

* [Dept of Mathematics & Statistics](https://www.dal.ca/faculty/science/math-stats/faculty-staff/our-faculty.html)

* [Dept of Biology](http://www.dal.ca/faculty/science/biology/faculty-staff/our-faculty/joe-bielawski.html)

<br>

## Arrival and Departure

Arrival: 27 May 2022 (Fri)

Depart:  07 June 2022 (Ts)

<br>

## PAML lab Materials

The lab exercises (PAML demo) have been published to a small website (link below). The site contains some additional resources that are worth a look when you have time. Please note that slides may change a little prior to the lab. I will post modified PDFs as required. 

{% include figure.html description="" url="/faculty/bielawski/img/MOLE_concept_map_v2.jpg" css="image-right" height="300px" %}

**PAML demo**: [PAML Lab website](https://awarnach.mathstat.dal.ca/~joeb/PAML_lab/lab.html)

**PAML demo resources**: [webpage](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/Resources.htm)

**PAML demo slides**: [2022 slides (PDF)](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/resources/pamlDEMO_MOLE2022_v2.pdf)

If you want doing the lab independently of the workshop (at home, on your own time and on your own computer), then you can do this by downloading all the necessary files from an archive [here](https://bitbucket.org/EvoWorks/protocol-paml-lab-at-mbl-workshop/downloads/), or you can download the files individually for each exercise as you need them [here](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/PAML_Lab_files.html).

**NOTE:**. *If you are doing the PAML Lab at the workshop, then use the VM and the 
symlink in your home directory named "moledata" to obtain the course data files*!!!

<br>

## Lecture Materials

**Current slides**

{% include figure.html description="" url="/faculty/bielawski/img/MolEvol.jpg" css="image-right" height="250px" %}

I am changing the lecture content for 2022. This lecture provides a more general background on evolutionary forces, and the Neutral and Nearly-Neutral theories of molecular evolution. Some details about fitting codon models to real data, have been moved to the "PAML Lab" lecture.  

* **2022 Lecture slides, Part 1:** [2022 coming soon]() 

* **2022 Lecture slides, Part 2:** [2022 coming soon]() 

<br>
**Some material from previous workshops**

I have included links to the 2019 slides below. This update includes more information on mechanistic processes of codon evolution (via the MutSel framework). Also, some might be interested in parts 3 and 4, which cover more advanced statistical topics, such as the requirements for likelihood inference and "phenomenological load" on parameter estimates. These topics will not be covered in 2022.

* **2019 Lecture slides, Parts 1 & 2**: [2019 PDF file1](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/slides/WH2019_Parts1and2_handouts.pdf)

* **2019 Lecture slides, Parts 3 & 4**: [2019 PDF file2](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/slides/WH2019_Parts3and4_handouts.pdf)


I updated the lecture slides on codon models for 2017. Because the older slides tend to have more details about fitting codon models to real data, I have included links to the 2015 and 2016 slides below; these slides provide more information about the powers and pitfalls of inference under codon models.

* **2016 Lecture slides, Part 1**: [2016 PDF file1](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/slides/WH2016_lecture_Parts1and2_FullPage.pdf)

* **2016 Lecture slides, Part 2**: [2016 PDF file2](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/slides/WH2016_lecture_Part3.pdf)

* **2015 Lecture slides, Part 1**: [2015 PDF file1](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/slides/Lecture_WH2015_Part1_2slidePage.pdf)

* **2015 Lecture slides, Part 2**: [2015 PDF file2](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/slides/ecture_WH2015_Part2_2slidePage.pdf)

<br>
**Key papers related to the lecture material:**


* Phenomenological load (PL) and biological conclusions under codon models:<br>
(_[Jones C.T., Youssef N., Susko E., Bielawski J.P., 2018. Phenomenological Load on Model Parameters Can Lead to False Biological Conclusions. Mol Biol Evol. 35(6):1473-1488.](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/resources/Jones_PL_MBE2018.pdf)_)

* Review of major inference challenges under codon models:<br>
(_[Jones C.T., Susko E., Bielawski J.P.,  2019. Looking for Darwin in genomic sequences: validity and success depends on the relationship between model and data. In Evolutionary Genomics: Statistical and Computational Methods. Maria Anisimova (ed.) 2nd edition, Human press.](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/resources/Jones_etal_2019_Chapter13.pdf)_)

* Positive selection, purifying selection, shifting balance & fitness landscapes:<br>
(_[Jones, C., Youssef, N., Susko, E. and Bielawski, J., 2017.  Shifting balance on a static mutation-selection landscape: a novel scenario of positive selection. Molecular Biology and Evolution, 34(2):391-407.](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/resources/ShiftBalance_MBE_2016.pdf)_)

* Improved inference of site-specific positive selection under a generalized parametric codon model when there are multi-nucleotide mutations and multiple nonsynonymous rates: <br>
(_[Dunn KA, Kenney T, Gu H, Bielawski JP. Improved inference of site-specific positive selection under a generalized parametric codon model when there are multinucleotide mutations and multiple nonsynonymous rates. BMC Evol Biol. 2019 Jan 14;19(1):22.](https://bmcevolbiol.biomedcentral.com/articles/10.1186/s12862-018-1326-7)_)


* ModL: restoring regularity when testing for positive selection: <br>
(_[Mingrone J, Susko E, Bielawski JP. ModL: exploring and restoring regularity when testing for positive selection. Bioinformatics. 2019 Aug 1;35(15):2545-2554.](https://academic.oup.com/bioinformatics/article/35/15/2545/5239994)_)

* Smoothed Bootstrap Aggregation (SBA) for assessing and correcting parameter estimate uncertainty in codon models:<br>
(_[Mingrone, J., Susko, E. and Bielawski, J., 2016. Smoothed bootstrap aggregation for assessing selection pressure at amino acid sites. Molecular Biology and Evolution, 33(11):2976-2989.](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/resources/Mingrone_SBA_MBE2016.pdf)_)

* Protocols, experimental design, and best practices for inference under complex codon models:<br>
(_[Bielawski, J.P., Baker, J.L. and Mingrone, J., 2016. Inference of episodic changes in natural selection acting on protein coding sequences via CODEML. Current Protocols in Bioinformatics, pp.6-15.](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/resources/UNIT6.15.pdf)_)

<br>

## Alternative Lab (advanced topics)

If you have some experience with codon models, and want to try out a tutorial for more advanced materials then use the link below to download an archive for a complete different set of PAML activities.  This tutorial focuses on detecting episodic protein evolution via **Branch-Site Model A**.  The tutorial also includes activities about (i) detecting **MLE instabilities**, (ii) carrying out **robustness analyses**, and (iii) use of **smoothed bootstrap aggregation (SBA)**.  The protocols for each activity are presented in Protocols in Bioinformatics UNIT 6.15.  The included PDF file for UNIT 6.16 also presents recommendations for "best practices" when carrying out a large-scale evolutionary survey for episodic adaptive evolution by using PAML.  The files required for this "alternative lab" are available via Bitbucket repository.  The repository link is given below.

**Advanced PAML demo:** [Bitbucket repository](https://bitbucket.org/EvoWorks/protocol-inference-of-episodic-selection/downloads)

**codeml_SBA:**  a program that implements Smoothed Bootstrap Aggregation (SBA) for assessing selection pressure at amino acid sites. [https://github.com/Jehops/codeml_sba](https://github.com/Jehops/codeml_sba)<br>

**DendroCypher:** a program to assist labelling the branches of a Newick-formatted tree-file for use with a “branch model” or a “branch-site codon model": [Bitbucket repository] (https://bitbucket.org/EvoWorks/dendrocypher/downloads/)

<br>


## "Best practices" in large-scale evolutionary surveys

Large-scale evolutionary surveys are now commonplace. But with the use of progressively more complex codon models, these surveys are fraught with perils. Complex models are more prone to statistical problems such as MLE irregularities, and some can be quite sensitive to model misspecification.  UNIT 6.16 (see above) provides some recommended "best practices" for a **2-phase approach** to quality control and robustness in evolutionary surveys.  We have applied these to a large scale survey for functional divergence in nuclear receptors during homing evolution, and we used experimental approaches to investigate hypotheses about the role of a particular nuclear receptor (NR2C1) as a key modulator of developmental pluripotnetiality during hominid evolution. The paper that illustrates the power of such an evolutionary surgery, and the importance of an experimental design having explicit protocols for "best practices", is given below.

**Example large-scale survey**: [PDF](http://awarnach.mathstat.dal.ca/~joeb/PAML_lab/resources/Baker.905.full.pdf)

<br>

## Alternative software for codon models in the ML framework

**HyPhy:**  comparative sequence analysis using stochastic evolutionary models; [http://www.hyphy.org/](http://www.hyphy.org/)

**DataMonkey:**  a server that supports a variety of HYPHY tools at no cost; [http://www.datamonkey.org/](http://www.datamonkey.org/)

**COLD:**  a program that implements a general-purpose parametric (GPP) codon model. Most codon models are special cases of the GPP codon model. [https://github.com/tjk23/COLD](https://github.com/tjk23/COLD)

**codeml_SBA:**  a program that implements smoothed Bootstrap Aggregation for Assessing Selection Pressure at Amino Acid Sites.[https://github.com/Jehops/codeml_sba](https://github.com/Jehops/codeml_sba)

**ModL:**  a program for restoring regularity when testing for positive selection using codon models [https://github.com/jehops/codeml_modl](https://github.com/jehops/codeml_modl)

{% include figure.html description="The Kidd" url="/faculty/bielawski/img/TheKidd1b.jpg" css="image-center" height="250px" %}
