---
layout: page
title: L. Lacey Knowles
permalink: /faculty-knowles/
---
{% include figure.html description="Lacey Knowles" url="faculty/knowles/img/LaceyPICT2-2021.jpg" css="image-right" height="250px" %}
### Institutional home page 

[lsa.umich.edu/eeb/people/faculty/knowlesl.html](https://lsa.umich.edu/eeb/people/faculty/knowlesl.html)

### Arrival and Departure 

May 31 to June 7

{% comment %}

### Lecture Materials

Lecture slides 2022
[Knowles-Woodshole-2022-Lect.pdf](https://github.com/molevolworkshop/molevolworkshop.github.io/files/8837808/Knowles-Woodshole-2022-Lect.pdf)
{% endcomment %}

### Demo and Tutorials
No demos in class, but if your are interested in spatial coalescent simulations and iDDC modeling check out:
https://becheler.github.io/quetzal-CRUMBS/index.html
and https://becheler.github.io/who-am-i/
This software package includes a test for the robustness of the MSC.

If you are interested in software for incorporating a speciation process into species delimitation, check out DELINEATE
https://github.com/jeetsukumaran/delineate

### Lecture in Brief

Capstone lecture: Evolutionary applications of genomic data 
[Knowles-Woodshole-2022-Lect.pdf](https://github.com/molevolworkshop/molevolworkshop.github.io/files/8837822/Knowles-Woodshole-2022-Lect.pdf)


I will highlight some common themes for model-based analyses with examples of a variety of evoltuionary applications of genomic data (e.g., phylogenetics, adaptive evolution, phylogeography, inferences about species boundaries): namely, 

- decisions/choices we make about modeul formulation
- recognizing the subjectivity of model formulation itself when making inferences
- decisions when applying to empirical data (e.g., how much data, what type of data, use of summary statistics etc)

#### Refereces
Here are some references for some papers I will hihglight as basckground reading.

#### Phylogeography

Becheler A, Knowles LL (2022) Computational resources for simulating under a spatial coalescent model across heterogeneous landscapes and testing hypotheses about the geography of genetic variation: QUETZAL-EGGS, -CRUMBS, -NEST and DECRYPT. Mol. Ecol. Res., in review. [QUEZTAL___MER__submitted_.pdf](https://github.com/molevolworkshop/molevolworkshop.github.io/files/8837803/QUEZTAL___MER__submitted_.pdf)

Carstens et al. (2017) Objective choice of phylogeographic models. Mol. Phylog. Evol. 116: 136-140.

He Q, Prado JR, Knowles LL (2017) [Inferring the geographic origin of a range expansion: latitudinal and longitudinal coordinates inferred from genomic data in an ABC framework with the program X-ORIGIN.](http://doi.org/10.1111/mec.14380) Mol. Ecol. 26:6908-6920.

Massatti R, Knowles LL (2016) Contrasting support for alternative models of genomic variation based on microhabitat preference: species-specific effects of climate change in alpine sedges. Mol. Ecol. 25:3974-3986. [Massatti&Knowles2016-ME.pdf](https://github.com/molevolworkshop/molevolworkshop.github.io/files/8837827/Massatti.Knowles2016-ME.pdf)


Oaks, J. R., J. Sukumaran, J. A. Esselstyn, C. W. Linkem, C. D. Siler, M., T. Holder, and R. M. Brown. 2013. Evidence for climate-driven diversification? A caution for interpreting ABC inferences of simultaneous historical events. Evolution 67:991â€“1010.

Papadopoulou A, Knowles LL (2016) Towards a paradigm shift in comparative phylogeography driven by trait-based hypotheses. Proc. Natl. Academy Sci. USA 113:8018-8024. [Papadopoulou&Knowles2016-PNAS.pdf](https://github.com/molevolworkshop/molevolworkshop.github.io/files/8837828/Papadopoulou.Knowles2016-PNAS.pdf)


#### Adadptive evolution

Decimated little brown bat population show potential for adaptive change. Scientific Reports. 10:3023. doi.org/10.1038/s41598-020-59797-4 [Auteri&Knowles2020.pdf](https://github.com/molevolworkshop/molevolworkshop.github.io/files/8837801/Auteri.Knowles2020.pdf) 

#### Species delimitation

Leache et al. (2018) [The spectre of too many species.](https://doi.org/10.1093/sysbio/syy051) Syst. Biol., in press. 

Jackson ND, Morales A, Carstens BC, O'Meara BC (2017) [Species delimitation with gene flow.](http://doi.org/10.1093/sysbio/syw117) Systematic Biology 66, 799-812. 

Sukumaran J, Knowles LL (2017) Multispecies coalescent delimits structure, not species. Proc. Natl. Academy Sci. USA 114:1607-1612.[Sukumaran&Knowles-PNAS-2017.pdf](https://github.com/molevolworkshop/molevolworkshop.github.io/files/8837832/Sukumaran.Knowles-PNAS-2017.pdf)

Sukumaran J, Holder MT, Knowles LL (2021) Incorporating the speciation process into species delimitation. PloS Comput. Biol.17(5):e1008924. https://doi.org/10.1371/journal.pcbi.1008924 [Sukumaran et al.2021-PlosCompBiol.pdf](https://github.com/molevolworkshop/molevolworkshop.github.io/files/8837833/Sukumaran.et.al.2021-PlosCompBiol.pdf)


Yang Z, Rannala B (2010) [Bayesian species delimitation using multilocus sequence data.](https://doi.org/10.1073/pnas.0913022107) PNAS 107:9264-9269.




